docs/proposals/interceptors.md

```diff
@@ -0,0 +1,196 @@
+
+## Description
+
+The Interceptor concept is much similar to the GraphQL middleware concept. The key attributes of interceptors are as follows.
+* Can execute any code before/after invoking a Resolver function
+* Can return errors
+* Have access to the `graphql:Context` and the `graphql:RequestInfo`
+* Can apply to multiple resolvers
+* Can pass state from one interceptor to another
+
+>NOTE: In the first phase, interceptors are supported only for `Query` and `Mutation` operations.
+
+### Interceptor Service Object
+The interceptor service object is the object that provides by the GrpahQL package. It includes a single remote function named `execute` that accepts `graphql:Context` and `graphql:Field` as the parameters. The function's return type is a union of `anydata` and `error`.
+
+```ballerina
+public type Interceptor distinct service object {
+    isolated remote function execute(Context context, Field 'field) returns anydata|error;
+};
+```
+
+### GraphQL Field Object
+GraphQL interceptor execute function accepts the `Field` object as an input parameter that consists of APIs to access the execution field information. Following is the implementation of the Field object.
+
+```ballerina
+public class Field {
+   public isolated function getName() returns string;
+ 
+   public isolated function getAlias() returns string;
+}
+```
+* The `getName()` function can be used to get the current execution field name.
+* The `getAlias()` function returns aliases if the current execution filed has aliases. If not, it returns the name of the field.
+
+Users will be able to access these APIs within the interceptor function.
+
+### GraphQL Interceptor
+Interceptors can be defined as a readonly service class that infers the Interceptor object provided by the GraphQL package. User-specific name can be used as the service class name. The Interceptor service class should have the implementation of the `execute()` remote function that infers from the interceptor service object. Apart from that, It can not have any other `resource/remote` methods inside the interceptor. Otherwise, it will be considered the invalid declaration of an interceptor function. However, the users are able to define the usual functions inside the interceptors. Code needed to be included in the interceptor should be kept inside the `execute()` function. To invoke the actual resolver function or the next interceptor the `context.resolve()` function can be called. When calling the `context.resolve()` function, the `graphql:Field` must be passed as the function argument. This will invoke the next interceptor if available. If not, it will call the actual resolver. Following is an example of a GraphQL interceptor.
```

Comment:
```suggestion
Interceptors can be defined as a readonly service class that infers the Interceptor object provided by the GraphQL package. User-specific name can be used as the service class name. The Interceptor service class should have the implementation of the `execute()` remote function that infers from the interceptor service object. Apart from that, It can not have any other `resource/remote` methods inside the interceptor. However, the users are able to define the usual functions inside the interceptors. Code needed to be included in the interceptor should be kept inside the `execute()` function. To invoke the actual resolver function or the next interceptor the `context.resolve()` function can be called. When calling the `context.resolve()` function, the `graphql:Field` must be passed as the function argument. This will invoke the next interceptor if available. If not, it will call the actual resolver. Following is an example of a GraphQL interceptor.
```